src/main/java/org/onedatashare/server/service/UploadService.java

```diff
@@ -32,29 +32,29 @@
 
     private static Map<UUID, LinkedBlockingQueue<Slice>> ongoingUploads = new HashMap<UUID, LinkedBlockingQueue<Slice>>();
 
-    public Mono<Integer> uploadChunk(String cookie, UUID uuid, Mono<FilePart> filePart, String credential,
+    public Mono<Boolean> uploadChunk(String cookie, UUID uuid, Mono<FilePart> filePart, String credential,
                                  String directoryPath, String fileName, Long totalFileSize, String googleDriveId, String idmap) {
         if (ongoingUploads.containsKey(uuid)) {
-            return sendFilePart(filePart, ongoingUploads.get(uuid));
-        }
-        else {
+            if(ongoingUploads.get(uuid).isEmpty())
+                return sendFilePart(filePart, ongoingUploads.get(uuid)).map(size -> true);
+            else return Mono.just(false);
```

Comment:
Few concerns on this -

1. What does this change do? Does it delay the chunk upload on the client side until true is returned? Asking since I did not find any JS change complimenting this backend change. I'm assuming FineUploader handles it automatically. Please confirm.

2. FineUploader has a retry count of 3 by default. If it doesn't receive 'true' in those retries, I am guessing the chunk will be lost. Can result in file corruption. Please correct me if I am wrong.

3. Current implementation doesn't make any use of the blocking queue. We can do away with it.

4. Instead of uploading 1 chunk at a time, can we use the queue (non-blocking) to store a few chunks (for example - queue size 3) and have concurrent uploading running during the RTT of sending the response to client and waiting for the next chunk?

5. Also, currently we are transmitting 500,000 bytes per chunk from client-size (which I guess we had set arbitrarily last sem for testing and never thought about its implications). Would reducing the chunk byte size to 256Kb result in a better performance?

Reply:
1.  FineUploader handles it automatically. If the chunk is not sent successfully, it will retry.
2. The upload will stop in that case. And since the drain never finishes, and the file will not be created. But you are right, front end parameter should be tweaked, and backend should keep a timer and discard the queue in that case. I will create that in the later pull request.
3. LinkedBlockingQueue is used in ClientUploadResource and ClientUploadSession. Can not do away with it.
4. Instead of that, we can increase the chunk size in the front end. It would be the same.
5. Reducing chunk byte size reduces transmit speed because of RTT of TCP. About chunk size should be set on the optimization of transfer speed. Right now we have not reached that stage of development yet.